File: mycollab-dao/src/main/java/com/mycollab/db/persistence/VelocityDriverDeclare.java
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.db.persistence;

import com.mycollab.core.MyCollabException;
import org.apache.ibatis.executor.parameter.ParameterHandler;
import org.apache.ibatis.mapping.BoundSql;
import org.apache.ibatis.mapping.MappedStatement;
import org.apache.ibatis.mapping.SqlSource;
import org.apache.ibatis.parsing.XNode;
import org.apache.ibatis.scripting.LanguageDriver;
import org.apache.ibatis.scripting.defaults.DefaultParameterHandler;
import org.apache.ibatis.session.Configuration;
import org.apache.ibatis.type.Alias;
import org.mybatis.scripting.velocity.SQLScriptSource;

import java.io.IOException;
import java.io.InputStream;

/**
 * @author MyCollab Ltd.
 * @since 4.0
 */
@Alias("velocity")
public class VelocityDriverDeclare implements LanguageDriver {
    private static final String TOTAL_COUNT_EXPR_MACRO;
    private static final String SELECT_EXPR_MACRO;

    static {
        TOTAL_COUNT_EXPR_MACRO = loadResource("totalCountExpr");
        SELECT_EXPR_MACRO = loadResource("selectExpr");
    }

    @Override
    public ParameterHandler createParameterHandler(MappedStatement mappedStatement, Object parameterObject, BoundSql boundSql) {
        return new DefaultParameterHandler(mappedStatement, parameterObject, boundSql);
    }

    @Override
    public SqlSource createSqlSource(Configuration configuration, XNode script, Class<?> parameterTypeClass) {
        if (parameterTypeClass == null) {
            parameterTypeClass = Object.class;
        }
        String templateStr = TOTAL_COUNT_EXPR_MACRO + " " + SELECT_EXPR_MACRO + " " + script.getNode().getTextContent();

        return new SQLScriptSource(configuration, templateStr, parameterTypeClass);
    }

    @Override
    public SqlSource createSqlSource(Configuration configuration, String script, Class<?> parameterTypeClass) {
        if (parameterTypeClass == null) {
            parameterTypeClass = Object.class;
        }
        String templateStr = new StringBuilder().append(SELECT_EXPR_MACRO)
                .append(" ").append(TOTAL_COUNT_EXPR_MACRO).append(" ")
                .append(script).toString();
        return new SQLScriptSource(configuration, templateStr,
                parameterTypeClass);
    }

    private static String loadResource(String id) {
        try (InputStream inputStream = VelocityDriverDeclare.class.getResourceAsStream(id)) {
            if (inputStream == null) {
                throw new MyCollabException("Can not load resource id " + id);
            } else {
                StringBuffer buf = new StringBuffer();
                int chBug;
                while ((chBug = inputStream.read()) != -1) {
                    buf.append((char) chBug);
                }

                return buf.toString();
            }
        } catch (IOException e) {
            throw new MyCollabException(e);
        }
    }
}

File: mycollab-dao/src/main/java/com/mycollab/db/persistence/VelocityDriverDeclare.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http:></http:>//www.gnu.org/licenses/>.
 */
package com.mycollab.db.persistence

import com.mycollab.core.MyCollabException
import org.apache.ibatis.executor.parameter.ParameterHandler
import org.apache.ibatis.mapping.BoundSql
import org.apache.ibatis.mapping.MappedStatement
import org.apache.ibatis.mapping.SqlSource
import org.apache.ibatis.parsing.XNode
import org.apache.ibatis.scripting.LanguageDriver
import org.apache.ibatis.scripting.defaults.DefaultParameterHandler
import org.apache.ibatis.session.Configuration
import org.apache.ibatis.type.Alias
import org.mybatis.scripting.velocity.SQLScriptSource
import java.io.IOException

/**
 * @author MyCollab Ltd.
 * @since 4.0
 */
@Alias("velocity")
class VelocityDriverDeclare : LanguageDriver {

    override fun createParameterHandler(mappedStatement: MappedStatement, parameterObject: Any, boundSql: BoundSql): ParameterHandler {
        return DefaultParameterHandler(mappedStatement, parameterObject, boundSql)
    }

    override fun createSqlSource(configuration: Configuration, script: XNode, parameterTypeClass: Class<*>?): SqlSource {
        var parameterTypeClass = parameterTypeClass
        if (parameterTypeClass == null) {
            parameterTypeClass = Any::class.java
        }
        val templateStr = "$TOTAL_COUNT_EXPR_MACRO $SELECT_EXPR_MACRO ${script.node.textContent}"

        return SQLScriptSource(configuration, templateStr, parameterTypeClass)
    }

    override fun createSqlSource(configuration: Configuration, script: String, parameterTypeClass: Class<*>?): SqlSource {
        var parameterTypeClass = parameterTypeClass
        if (parameterTypeClass == null) {
            parameterTypeClass = Any::class.java
        }
        val templateStr = "$SELECT_EXPR_MACRO $TOTAL_COUNT_EXPR_MACRO $script"
        return SQLScriptSource(configuration, templateStr, parameterTypeClass)
    }

    companion object {
        private val TOTAL_COUNT_EXPR_MACRO: String
        private val SELECT_EXPR_MACRO: String

        init {
            TOTAL_COUNT_EXPR_MACRO = loadResource("totalCountExpr")
            SELECT_EXPR_MACRO = loadResource("selectExpr")
        }

        private fun loadResource(id: String): String {
            try {
                VelocityDriverDeclare::class.java.getResourceAsStream(id).use { inputStream ->
                    if (inputStream == null) {
                        throw MyCollabException("Can not load resource id $id")
                    } else {
                        val buf = StringBuffer()
                        var chBug = inputStream.read()
                        while (chBug != -1) {
                            buf.append(chBug.toChar())
                            chBug = inputStream.read()
                        }

                        return buf.toString()
                    }
                }
            } catch (e: IOException) {
                throw MyCollabException(e)
            }

        }
    }
}